### PR TITLE
recursion depth

### PR DESCRIPTION
**Pull request by jordens**
- update the generic bound logic in the derive macro

This rewrites the type bound logic after having analyzed it again and
after the changes due to adding the cont generic in `Miniconf`.

The Miniconf<Y> const generic Y is now the miniconf recursion depth
again. And it's an upper bound to the state size.
Before, the idea of making it exactly the key length war broken:
We had `impl Miniconf<0> for Option<T> where T: SerDe`
but for `#[miniconf(defer(1))] a: [Option<T>; N]` it would not use the
miniconf API of `Option<T>` as desired as it would for
`#[miniconf(defer(0))] a: Option<T>`.
Using the recursion depth as indicator which API to call fixes this.

At the cost of the state size now being potentially a bit too large.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -141,8 +141,8 @@ Homogeneous [core::array]s can be made accessible either
 
 `Option` is used
 1. like a standard `serde` Option, or
-2. with `#[miniconf(defer(0))]` to support a value that may be absent (masked) at runtime.
-3. with `#[miniconf(defer(D))]` and `D >= 1` to support mask sub-trees at runtime.
+2. with `#[miniconf(defer(1))]` to support a value that may be absent (masked) at runtime.
```
Comment (ryan-summers) [inline README.md:144]: Are `defer` and `defer(1)` equivalent?

Comment (ryan-summers) [reply README.md:144]: Looks like yes, can we note that?

File: src/lib.rs [hunk 2]
```diff
@@ -149,7 +149,23 @@ impl Key for &str {
...
 /// Trait exposing serialization/deserialization of nodes by keys/paths and traversal.
 ///
-/// The depth parameter `Y` is the maximum key length: the depth/height of the tree.
+/// The depth parameter `Y` is the maximum miniconf recursion depth.
+/// An implementor chooses a `Z <= Y` and may then in the recursive methods (directly
+/// and indirectly) retrieve at most `Z` keys from the `keys` and `indices` iterators
+/// and call the `func` callback at most `Z` times. It may
+/// use at most `Miniconf<Z>` on its inner types. `Z` must be consistent, i.e. if
+/// an implementor retrieves `X` keys before recursing to an inner type, it may at
+/// most use `Miniconf<Y - X>` on its inner types.
```
Comment (ryan-summers) [inline src/lib.rs:158]: I have absolutely no idea what this means... Is there a  way we can state this in simpler terms?

File: src/lib.rs [hunk 3]
```diff
@@ -149,7 +149,23 @@ impl Key for &str {
...
+///
+/// Furthermore the `derive(Miniconf)` macro assumes that any implementor of `Miniconf<Y>`
+/// will require its generic types to be exactly `Miniconf<Y - 1>`. This condition is upheld
+/// by all implementations in this crate. It is only violated
+/// when deriving `Miniconf` for a struct that (a) forwards its own type parameters as type
+/// parameters to its field types, (b) uses `Miniconf` on those fields, and (c) those field
+/// types use their type parameters at other levels than `Miniconf<Y - 1>`.
```
Comment (ryan-summers) [inline src/lib.rs:165]: I don't particularly understand why this matters...? Is this something that needs to be upheld, or is this purely informational about the design? Maybe we can move it under a `Design` subheading or something?

File: tests/generics.rs [hunk 4]
```diff
@@ -86,6 +90,17 @@ fn generic_atomic() {
...
+    assert_eq!(metadata.max_length, "/opt1/0/0".len());
+}
+
+#[test]
+fn test_failure() {
+    type A<T> = [[T; 0]; 0];
+    #[derive(Miniconf)]
+    struct S<T: serde::Serialize + serde::de::DeserializeOwned>(
+        // this wrongly infers T: Miniconf<1> instead of T: SerDe
+        // adding the missing bound is a workaround
```
Comment (ryan-summers) [inline tests/generics.rs:103]: What is this actually testing?

File: tests/option.rs [hunk 5]
```diff
@@ -155,3 +155,16 @@ fn option_absent() {
...
+fn array_option() {
+    #[derive(Copy, Clone, Default, Miniconf)]
+    struct S {
+        #[miniconf(defer(1))]
+        a: Option<u32>,
+        #[miniconf(defer(1))]
+        b: [Option<u32>; 1],
+        #[miniconf(defer(2))]
+        c: [Option<u32>; 1],
+    }
```
Comment (ryan-summers) [inline tests/option.rs:169]: I don't see any actual tests here

File: src/lib.rs [hunk 6]
```diff
@@ -149,7 +149,23 @@ impl Key for &str {
 
 /// Trait exposing serialization/deserialization of nodes by keys/paths and traversal.
 ///
-/// The depth parameter `Y` is the maximum key length: the depth/height of the tree.
+/// The depth parameter `Y` is the maximum miniconf recursion depth.
+/// An implementor chooses a `Z <= Y` and may then in the recursive methods (directly
+/// and indirectly) retrieve at most `Z` keys from the `keys` and `indices` iterators
+/// and call the `func` callback at most `Z` times. It may
```
Comment (ryan-summers) [inline src/lib.rs:155]: I don't understand these docs - what `keys` and `indices` are they referring to? Seems like `indices` is referring to the member on this trait, but `keys` might be an arg provided? In general, I'm having a hard time understanding these docs and what they're supposed to mean

**Review by ryan-summers — CHANGES_REQUESTED** [review]
I'm very thoroughly lost on what the depth parameters are supposed to be

File: tests/generics.rs (hunk 4 above)
Comment (jordens) [reply tests/generics.rs:103]: Added comments.

File: tests/option.rs (hunk 5 above)
Comment (jordens) [reply tests/option.rs:169]: Added comment.

File: README.md (hunk 1 above)
Comment (jordens) [reply README.md:144]: Added docs.

File: src/lib.rs (hunk 6 above)
Comment (jordens) [reply src/lib.rs:155]: Clarified the docs. Please check again.

File: src/lib.rs (hunk 2 above)
Comment (jordens) [reply src/lib.rs:158]: See above. If you understand, let me know if you see a simpler way.

File: src/lib.rs (hunk 3 above)
Comment (jordens) [reply src/lib.rs:165]: It matters and it's a limitation (always was). See the tests and the added comments there.
I'll add a link from the readme here.

**Comment by jordens** [comment]
> I'm very thoroughly lost on what the depth parameters are supposed to be

What specific questions do you have that are not answered by the docs?

File: src/lib.rs [hunk 7]
```diff
@@ -149,23 +149,92 @@ impl Key for &str {
...
+/// 1. as a leaf like a standard `serde` Option, or
+/// 2. with `#[miniconf(defer(1))]` to support a leaf value that may be absent (masked) at runtime.
+/// 3. with `#[miniconf(defer(D))]` and `D >= 2` to support masking sub-trees at runtime.
+///
+/// ## Bounds on generics
+///
+/// The macro adds bounds to generic types of the struct it is acting on.
+/// E.g. If a generic type parameter `T` of the struct `S<T>`is used as a type parameter to a
+/// field type `a: F1<F2<T>>` the type `T` will be considered to reside at depth `X = 2` (as it is
+/// within `F2` which is within `F1`) and the following bounds will applied:
```
Comment (ryan-summers) [inline src/lib.rs:206]:
```suggestion
/// within `F2` which is within `F1`) and the following bounds will be applied:
```

File: src/lib.rs [hunk 8]
```diff
@@ -149,23 +149,92 @@ impl Key for &str {
...
+///     #[miniconf(defer)]
+///     data: [u32; 2],
+/// }
+/// #[derive(Miniconf)]
+/// struct Settings {
+///     // Accessed with path `nested/data/0` or `nested/data/1`
+///     #[miniconf(defer(2))]
+///     nested: Nested,
+///
+///     // Accessed with path `external`
```
Comment (ryan-summers) [inline src/lib.rs:234]:
```suggestion
///     // Accessed with path `/nested/data/0` or `/nested/data/1`
///     #[miniconf(defer(2))]
///     nested: Nested,
///
///     // Accessed with path `/external`
```

Don't we enforce leading prefixes now?

File: src/lib.rs [hunk 9]
```diff
@@ -149,23 +149,92 @@ impl Key for &str {
...
+/// `Option` is used
+/// 1. as a leaf like a standard `serde` Option, or
+/// 2. with `#[miniconf(defer(1))]` to support a leaf value that may be absent (masked) at runtime.
+/// 3. with `#[miniconf(defer(D))]` and `D >= 2` to support masking sub-trees at runtime.
+///
+/// ## Bounds on generics
+///
+/// The macro adds bounds to generic types of the struct it is acting on.
+/// E.g. If a generic type parameter `T` of the struct `S<T>`is used as a type parameter to a
+/// field type `a: F1<F2<T>>` the type `T` will be considered to reside at depth `X = 2` (as it is
```
Comment (ryan-summers) [inline src/lib.rs:205]: Can we just write this out as:
```rs
type F1<T> = ...;
type F2<T> = ...;

struct S<T> {
    a: F1<F2<T>>,
}
```

It actually took me a couple read throughs to realize this is what it was trying to describe.

File: src/lib.rs [hunk 10]
```diff
@@ -149,23 +149,92 @@ impl Key for &str {
...
 /// Trait exposing serialization/deserialization of nodes by keys/paths and traversal.
 ///
-/// The depth parameter `Y` is the maximum miniconf recursion depth.
-/// An implementor chooses a `Z <= Y` and may then in the recursive methods (directly
-/// and indirectly) retrieve at most `Z` keys from the `keys` and `indices` iterators
-/// and call the `func` callback at most `Z` times. It may
-/// use at most `Miniconf<Z>` on its inner types. `Z` must be consistent, i.e. if
-/// an implementor retrieves `X` keys before recursing to an inner type, it may at
-/// most use `Miniconf<Y - X>` on its inner types.
+/// The keys used to locate nodes can bei either iterators over `usize` or iterators
```
Comment (ryan-summers) [inline src/lib.rs:152]:
```suggestion
/// The keys used to locate nodes can be either iterators over `usize` or iterators
```

**Review by ryan-summers — CHANGES_REQUESTED** [review]
The docs look much more understandable, thanks for taking the time to rework them

File: src/lib.rs (hunk 8 above)
Comment (jordens) [reply src/lib.rs:234]: Everything before the first / is ignored. This only applies in json_core
Added doc about that there.

File: src/lib.rs (hunk 9 above)
Comment (jordens) [reply src/lib.rs:205]: Added doctest.

File: src/lib.rs (hunk 7 above)
Comment (jordens) [reply src/lib.rs:206]: Ack

**Comment by ryan-summers** [comment]
@jordens Once you update this, we can probably merge

File: src/lib.rs [hunk 11]
```diff
@@ -377,6 +486,16 @@ pub trait Miniconf<const Y: usize = 1> {
     ///
     /// See also [Miniconf::iter_paths].
     ///
+    /// ```
+    /// # use miniconf::{Miniconf, Metadata};
+    /// # use heapless::String;
+    /// #[derive(Miniconf)]
+    /// struct S {foo: u32};
```
Comment (ryan-summers) [inline src/lib.rs:503]: Gotta remove teh `heapless::String` throughout or fix the import

Comment (jordens) [reply src/lib.rs:504]: It's both needed and correct. The problem is elsewhere.

**Review by ryan-summers — APPROVED** [review]
